### PR TITLE
Refactor splice interpreter

### PR DESCRIPTION
**Pull request by nicolasstucki**
Extract the general purpose logic from the splice interpreter. The new interpreter class will be the basis for the macro annotation interpreter. The splice interpreter only keeps logic related with level -1 quote and type evaluation.

Part of https://github.com/dotty-staging/dotty/tree/design-macro-annotations

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/quoted/Interpreter.scala [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+import dotty.tools.dotc.core.TypeErasure
+import dotty.tools.dotc.core.Types._
+import dotty.tools.dotc.quoted._
+import dotty.tools.dotc.transform.TreeMapWithStages._
+import dotty.tools.dotc.typer.ImportInfo.withRootImports
+import dotty.tools.dotc.util.SrcPos
+import dotty.tools.repl.AbstractFileClassLoader
+
+
+/** List of classes of the parameters of the signature of `sym` */
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/quoted/Interpreter.scala:34]: That comment seems misplaced? It also appears above `def paramsSig`

File: compiler/src/dotty/tools/dotc/quoted/Interpreter.scala [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+abstract class Interpreter(pos: SrcPos, classLoader: ClassLoader)(using Context):
+  import Interpreter._
+  import tpd._
+
+  type Env = Map[Symbol, Object]
+
+  /** Returns the interpreted result of interpreting the code a call to the symbol with default arguments.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/quoted/Interpreter.scala:41]: Looks like this comment was here before but I'm having trouble interpreting it myself :). Is there a word missing around "the code a call",  should it be "the code of a call" ?

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/quoted/Interpreter.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/Interpreter.scala]: Removed. I must have pasted it by mistake.

File: compiler/src/dotty/tools/dotc/quoted/Interpreter.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/quoted/Interpreter.scala]: I rewrote the sentence.

**Review by smarter — APPROVED** [review]

**Pull request merged**